modules/graphics/src/mrb_spritesheet.cxx

```diff
@@ -125,60 +168,60 @@ spritesheet_render(mrb_state *mrb, mrb_value self)
   mrb_int index;
   mrb_float x, y, z;
   mrb_value options = mrb_nil_value();
-  RenderState render_op;
+  RenderState render_state;
   mrb_get_args(mrb, "fffi|H", &x, &y, &z, &index, &options);
   if (!mrb_nil_p(options)) {
-    mrb_value keys = mrb_hash_keys(mrb, options);
-    int len = mrb_ary_len(mrb, keys);
-    const mrb_value *keys_ary = RARRAY_PTR(keys);
-
-    for (int i = 0; i < len; ++i) {
-      mrb_value key = keys_ary[i];
-
-      if (mrb_symbol_p(key)) {
-        mrb_value val = mrb_hash_get(mrb, options, key);
-        // :opacity
-        if (mrb_symbol(key) == id_opacity) {
-          render_op.opacity = mrb_to_flo(mrb, val);
-
-        // :color
-        } else if (mrb_symbol(key) == id_color) {
-          render_op.color = mmrb_to_vector4(mrb, val);
-
-        // :tone
-        } else if (mrb_symbol(key) == id_tone) {
-          render_op.tone = mmrb_to_vector4(mrb, val);
-
-        // :ox
-        } else if (mrb_symbol(key) == id_ox) {
-          render_op.origin.x = mrb_to_flo(mrb, val);
-
-        // :oy
-        } else if (mrb_symbol(key) == id_oy) {
-          render_op.origin.y = mrb_to_flo(mrb, val);
-
-        // :angle
-        } else if (mrb_symbol(key) == id_angle) {
-          render_op.angle = mrb_to_flo(mrb, val);
-
-        // :transform
-        } else if (mrb_symbol(key) == id_transform) {
-          render_op.transform = mmrb_to_transform(mrb, val);
-        }
-      }
-    }
+    set_render_options(mrb, options, &render_state);
   }
-  render(mrb, self, glm::vec3(x, y, z), index, render_op);
+  render(mrb, self, glm::vec3(x, y, z), index, render_state);
+  return self;
+}
+
+static mrb_value
+spritesheet_push_quad(mrb_state *mrb, mrb_value self)
+{
+  RenderState render_state;
+  Moon::VertexBuffer *dest_vbo;
+  Moon::VertexBuffer *vbo;
+  mrb_float x, y, z;
+  mrb_value options = mrb_nil_value();
+  mrb_int offset = 0;
+  GLuint index;
+  Moon::Vertex vertices[4];
+  GLuint indices[] = { 0, 1, 3, 2, 3, 1 };
+  int i;
+  mrb_get_args(mrb, "dfffi|H",
+    &dest_vbo, &vbo_data_type,
+    &x, &y, &z, &offset,
+    &options
+  );
+  if (!mrb_nil_p(options)) {
+    set_render_options(mrb, options, &render_state);
+  }
+
+  vbo = get_vbo(mrb, moon_mrb_iv_get_no_nil(mrb, self, KEY_VBO));
+  Moon::Vector2 pos(x, y);
+  render_state.opacity = 1.0;
+  index = offset * 4;
+  // now apply render state changes
+  for (int i = 0; i < 4; ++i) {
+    vertices[i] = vbo->GetVertex(index + i);
```

Comment:
Why are you doing a lookup here? You're adding a new quad to the VBO, so there should be no lookups going on..